proposals/buildfile.md

```diff
@@ -0,0 +1,186 @@
+            - "**/goose.class"
+            - "**/moose.class"
+          includes:
+            - "**/*.class"
+
+      # another layer, only globally defined file permissions are applied here
+    - name: "other"
+      files:
+        - from: "build/other"
+          to: "/app"
+
+      # a archive layer using a tar, default mediaType
+    - name: "some tar layer"
+      archive: "build/generated.tar"
+
+      # a foreign layer using the optional mediatype for archive layers
+    - name: "some foreign layer"
+      mediaType: "application/vnd.docker.image.rootfs.foreign.diff.tar.gzip"
+      archive: "https://somewhere.com/layer"
+      # should we include size and digest here? I guess we can always ad tings
+```
+
+## Layers
+
+`layers.entries` are a list of `layer` directives
+
+`layer` directives can be `archive` or `file` layers
+
+a `archive` layer consists of 3 parts
+* `name`: the name/description of the layer (metadata)
+* `archive`: a tar file to include as a layer
+* `mediatype`: the mediatype of archive
+
+a `file` layer consists of 3 parts
+* `name`: the name of the layer (metadata)
+* `properties`: on disk metadata for all files in the layer
+* `files`: a list of *copy directies* (see below)
+
+`files` are a list of *copy directives* that consist of 2 required and 3 optional parts
+* `to` *required*: an absolute path to copy the files to on the container
+* `from` *required*: a file, or directory on disk to copy from
+* `includes`: only includes the patterns matched in this parameter
+* `exludes`: excludes all files (higher prescendence than `includes`) matched in this parameter
+* `properties`: FilesProperties to represent on disk metadata for all files in this copy directive
+
+### FileProperties
+
+A list of properties that can be user specified for each file in a file layer
+* filePermissions: An octal representing permissions for files
+* directoryPermissions: An octal representing permissions for directories
+* user: The ownership user property
+* group: The ownership group property
+* timestamp: millis since epoch or iso8601 creation time
+
+File properties are available at 3 levels
+* Global (`layers.properties`): applies to all layers
+* Layer (`layers.<entry>.properties`): applies to a single layer
+* Copy (`layers.<entry>.<copy>.properties`): applies to a single copy directive
+
+Each property (`filePermissions`, `directoryPermissions`, etc) can be defined at any level and are resolved in the follow order
+- All properties in `Copy` are applied first
+- Any properties not in `Copy` are applied from `Layer`
+- Any properties not in `Copy` or `Layer` are applied from `Global`
+- Any properties not defined anywhere use jib system defaults.
+
+### Base image value inheritance
+The value(s) defined in the base image are preserved and propogated into the
+config of the new container.
+
+The behavior of the buildfile values post-inheritance must be considered
+
+These parameters will allow appending to the base image value:
```

Comment:
Separate out `environment` and `labels`: they will override values with the same key.

Reply:
ah good call.